### PR TITLE
Issue #10228: Fix Intermittent AutofillCreditCardsAddressesStorageTest

### LINKED ISSUE #10228: Intermittent services-sync-autofill - mozilla.components.service.sync.autofill.AutofillCreditCardsAddressesStorageTest - delete credit card
#10140 landed directly before this failure and touched the relevant code. @gabrielluong , please check the test failure.

### Firebase Test Run:
https://firefoxci.taskcluster-artifacts.net/Nc06EzrcT4CVr2CTyzb9OQ/0/public/logs/live_backing.log
### Stacktrace:
```

[task 2021-05-06T02:11:38.296Z] SUITE: mozilla.components.service.sync.autofill.AutofillCreditCardsAddressesStorageTest
[task 2021-05-06T02:11:38.297Z]   TEST: delete credit card
[task 2021-05-06T02:11:38.297Z]     [Robolectric] mozilla.components.service.sync.autofill.AutofillCreditCardsAddressesStorageTest.delete credit card: sdk=28; resources=binary
[task 2021-05-06T02:11:38.394Z]     I/MonitoringInstr: Instrumentation started!
[task 2021-05-06T02:11:38.495Z]   FAILURE
[task 2021-05-06T02:11:38.495Z] 
[task 2021-05-06T02:11:38.495Z] java.lang.AssertionError: expected:<1> but was:<2>
[task 2021-05-06T02:11:38.495Z] 	at org.junit.Assert.fail(Assert.java:88)
[task 2021-05-06T02:11:38.495Z] 	at org.junit.Assert.failNotEquals(Assert.java:834)
[task 2021-05-06T02:11:38.495Z] 	at org.junit.Assert.assertEquals(Assert.java:118)
[task 2021-05-06T02:11:38.495Z] 	at org.junit.Assert.assertEquals(Assert.java:144)
[task 2021-05-06T02:11:38.496Z] 	at mozilla.components.service.sync.autofill.AutofillCreditCardsAddressesStorageTest$delete credit card$1.invokeSuspend(AutofillCreditCardsAddressesStorageTest.kt:200)
[task 2021-05-06T02:11:38.496Z] 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
[task 2021-05-06T02:11:38.496Z] 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
[task 2021-05-06T02:11:38.496Z] 	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
[task 2021-05-06T02:11:38.496Z] 	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
[task 2021-05-06T02:11:38.496Z] 	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
[task 2021-05-06T02:11:38.496Z] 	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
[task 2021-05-06T02:11:38.496Z] 	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
[task 2021-05-06T02:11:38.496Z] 	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
[task 2021-05-06T02:11:38.496Z] 	at mozilla.components.service.sync.autofill.AutofillCreditCardsAddressesStorageTest.delete credit card(AutofillCreditCardsAddressesStorageTest.kt:187)
[task 2021-05-06T02:11:38.497Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[task 2021-05-06T02:11:38.497Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[task 2021-05-06T02:11:38.497Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2021-05-06T02:11:38.497Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[task 2021-05-06T02:11:38.497Z] 	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:601)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
[task 2021-05-06T02:11:38.497Z] 	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:260)
[task 2021-05-06T02:11:38.497Z] 	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
[task 2021-05-06T02:11:38.497Z] 	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[task 2021-05-06T02:11:38.497Z] 	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
[task 2021-05-06T02:11:38.497Z] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[task 2021-05-06T02:11:38.497Z] 	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:154)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
[task 2021-05-06T02:11:38.497Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[task 2021-05-06T02:11:38.497Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[task 2021-05-06T02:11:38.497Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2021-05-06T02:11:38.497Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
[task 2021-05-06T02:11:38.497Z] 	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
[task 2021-05-06T02:11:38.497Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[task 2021-05-06T02:11:38.497Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[task 2021-05-06T02:11:38.497Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2021-05-06T02:11:38.497Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
[task 2021-05-06T02:11:38.497Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[task 2021-05-06T02:11:38.497Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[task 2021-05-06T02:11:38.497Z] 	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
[task 2021-05-06T02:11:38.497Z] 	at java.lang.Thread.run(Thread.java:748)
```
### Build: 
https://github.com/mozilla-mobile/android-components/commit/867def8a02e7b84155ced90f87f71e2c71761e0b

### PR DESCRIPTION
**Pull request by gabrielluong**
Fixes #10228 

- To get around the intermittent, we call `getCreditCard(creditCard.guid)` and make sure it's gone instead of calling `getAllCreditCards()`
- Change `getCreditCard` and `getAddress` to be able to return null when no record can be found

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by gabrielluong** [comment]
@grigoryk I still see failure with this. Maybe we need to visit the db transactions?

File: components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/AutofillCreditCardsAddressesStorage.kt [hunk 1]
```diff
@@ -136,8 +146,16 @@ class AutofillCreditCardsAddressesStorage(
             conn.getStorage().addAddress(addressFields.into()).into()
         }
 
-    override suspend fun getAddress(guid: String): Address = withContext(coroutineContext) {
-        conn.getStorage().getAddress(guid).into()
+    /**
+     * @throws [NoSuchRecord] if the address does noot exist.
+     */
+    @Throws(NoSuchRecord::class)
```
Comment (csadilek) [inline components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/AutofillCreditCardsAddressesStorage.kt:152]: ^^ This throws is not needed. Method doesn't throw.

**Review by csadilek — APPROVED** [review]
Thanks, one nit below. We can file a follow-up for the setup here e.g. do we need better TX support? Or should we configure one storage / per test instead?

Either way, good to get the intermittents fixed fir.st

File: components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/AutofillCreditCardsAddressesStorage.kt [hunk 2]
```diff
@@ -115,8 +117,16 @@ class AutofillCreditCardsAddressesStorage(
         conn.getStorage().updateCreditCard(guid, updatableCreditCardFields.into())
     }
 
-    override suspend fun getCreditCard(guid: String): CreditCard = withContext(coroutineContext) {
-        conn.getStorage().getCreditCard(guid).into()
+    /**
+     * @throws [NoSuchRecord] if the credit card does not exist.
+     */
+    @Throws(NoSuchRecord::class)
```
Comment (csadilek) [inline components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/AutofillCreditCardsAddressesStorage.kt:123]: ^^ Same, don't need `@Throws` here.

**Pull request merged**